_posts/2018-05-31-regular-types.md

```diff
@@ -0,0 +1,874 @@
+definition / usage / discussion of Regular. I believe that the missing piece for
+Regular is "thread compatible" in the general case (although there are other
+valid options), and existing design precedent in the library already follows
+that direction. Builtins and our vocabulary types are both Regular and
+thread-compatible.
+
+## Flavors of Race-Free, plus Regular
+
+While Regular+thread-compatible is the most common (and most like `int`)
+combination that actually describes good/built-in-like types, other options may
+work in some situations. These options correspond roughly to the answers for
+"How do you know that operations on this instance do not cause data races?"
+
+For any given instance of a type, you might know one of several possible things
+that allow you to operate on it race-free.
+
+-   **Thread-compatible** and not shared with other threads for writing. If
+    you've been handed a (non-racing) `const T&` you can operate on this in
+    const fashion. If necessary, you can copy it to ensure there are no lurking
+    references and perform any computation / mutation safely (but
+    inefficiently). With minor knowledge (the instance isn't shared), a `T&` can
+    be used safely as if it were `T`.
+-   It has **dependent-preconditions**, but for a particular instance + any
+    dependent data, the program structure guarantees safe usage.
+-   **Single-threaded usage** - There is only one thread in the program and thus
+    all instances of the type are safe to use, or a given instance is known to
+    not be shared among threads.
+
+P0898r2's [concepts.lib.general.equality] p3, cited above, would fix the
+standard library's stance on Regular to be entirely the first option ... at the
+cost of pointers not being considered Regular because of semantic requirements
+and implementation-defined behavior on comparison. (If deletion is "as-if all
+other pointers to this object are scribbled over" that certainly violates the
+restriction on the value of the object changing out from under us.)
+
+All three of these options provide points in the type design space that are
+useful in conjunction with the existing definitions for Regular.
+
+-   **Thread-compatible** + Regular is what we really want for user-defined
+    types that mimic built-ins. This lets us reason about an instance in the
+    expected fashion and use it efficiently in conjunction with generic
+    algorithms. Types that have mutable data may have some overhead to support
+    this.
+-   **Dependent-preconditions** with knowledge that an instance + its dependent
+    data are safe to use. This is the common usage for `string_view` when we use
+    it as a non-owning parameter type: the underlying buffer will outlive the
+    function call and is immutable for the duration of the call. Given that
+    external knowledge of that underlying buffer, `string_view` behaves as if it
+    were Regular.
```

Comment:
Optionally add: "This makes sense, given that `string_view` was designed to be a drop-in replacement for `const string&`, and although references are not Regular, `string` is.

Reply:
Done, with some grammar changes.